ui/src/saros/ui/browser/SWTBrowserWrapper.java

```diff
@@ -0,0 +1,192 @@
+package saros.ui.browser;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicReference;
+import org.eclipse.swt.browser.Browser;
+import org.eclipse.swt.browser.BrowserFunction;
+import org.eclipse.swt.browser.ProgressEvent;
+import org.eclipse.swt.browser.ProgressListener;
+import org.eclipse.swt.events.DisposeEvent;
+import org.eclipse.swt.widgets.Composite;
+import org.eclipse.swt.widgets.Display;
+
+public class SWTBrowserWrapper implements IBrowserWrapper {
+
+  private enum ReadyState {
+    LOADING,
+    COMPLETE
+  }
+
+  private final Browser browser;
+  private ReadyState state = ReadyState.LOADING;
+  private List<Runnable> runOnDisposalList = new ArrayList<Runnable>();
+  private final ProgressListener readyStateListener =
+      new ProgressListener() {
+        @Override
+        public void completed(ProgressEvent event) {
+          state = ReadyState.COMPLETE;
+        }
+
+        @Override
+        public void changed(ProgressEvent event) {
+          state = ReadyState.LOADING;
+        }
+      };
+
+  public SWTBrowserWrapper(Composite parent, int style) {
+    browser = new Browser(parent, style);
+
+    browser.addDisposeListener(
+        (DisposeEvent e) -> {
+          for (Runnable runnable : runOnDisposalList) {
+            runnable.run();
+          }
+        });
+  }
+
+  @Override
+  public boolean setFocus() {
+    return this.syncExec(
+        () -> {
+          return browser.setFocus();
+        });
+  }
+
+  @Override
+  public void setSize(final int width, final int height) {
+    this.syncExec(() -> browser.setSize(width, height));
+  }
+
+  @Override
+  public String getUrl() {
+    return this.syncExec(
+        () -> {
+          return browser.getUrl();
+        });
+  }
+
+  @Override
+  public boolean loadUrl(final String url, final long timeout) {
+    browser.getShell().open();
+
+    this.state = ReadyState.LOADING;
+    browser.addProgressListener(readyStateListener);
+
+    browser.setUrl(url);
+    boolean successful = this.waitForPageLoaded(url, timeout);
+
+    browser.removeProgressListener(readyStateListener);
+
+    return successful;
+  }
+
+  /**
+   * Waits for a page to be loaded.
+   *
+   * @param url
+   * @param timeout
+   * @return
+   */
+  private boolean waitForPageLoaded(final String url, final long timeout) {
+
+    Runnable waitForLoaded =
+        () -> {
+          while (this.state != ReadyState.COMPLETE) {
+            if (!browser.getDisplay().readAndDispatch()) {
```

Comment:
I do not think that this is working at all. Do you even know what you are doing here ? You are trying to dispatch UI events from another thread. I guess that this thread always crashes due to an illegal access exception.

Reply:
It seems to work, but I know what you mean. I would expect a `SWTException: Invalid thread access`. I already had different versions. Which would you prefer. I am not sure which is better:
* Just wrap the if-block it into a `syncExec`
* Remove the display operations and use a Semaphore (initial permits = 0) which blocks in `acquire` here until it is released in the ProgressListener.

Reply:
We have to clarify what thread invoked the listener. If this is always not the SWT thread then simply wait (with a timeout) for the result. I would also recommend to block here even it is called from the SWT because if you call readAndDispatch you are may expect that setUrl can be called again. I do not know how long it takes to load the pages but a small UI freeze might be acceptable in this situation.

Reply:
The execution thread of the listeners seems to depend on the implementation:
* IE: It seems like the listeners are executed [directly](https://github.com/eclipse/eclipse.platform.swt/blob/6bdf4f12c125b636189a4de55260df5a71775f2e/bundles/org.eclipse.swt/Eclipse%20SWT%20Browser/win32/org/eclipse/swt/browser/IE.java#L656)
* GTK-WebKit1: [it is executed directly](https://github.com/eclipse/eclipse.platform.swt/blob/6bdf4f12c125b636189a4de55260df5a71775f2e/bundles/org.eclipse.swt/Eclipse%20SWT%20WebKit/gtk/org/eclipse/swt/browser/WebKit.java#L2464),  
GTK-WebKit2: it is executed in the ui thread 
Cocoa-WebKit: it is [executed directly](https://github.com/eclipse/eclipse.platform.swt/blob/6bdf4f12c125b636189a4de55260df5a71775f2e/bundles/org.eclipse.swt/Eclipse%20SWT%20WebKit/cocoa/org/eclipse/swt/browser/WebKit.java#L821)

Reply:
It seems that this is always called in the SWT EDT. So if you make sure waitForPageLoaded is invoked in the EDT (seems loadUrl is the only method that is not executed on the SWT) you can simply spin wait via.

    while (!completed)
      if (!display.readAndDispatch()) Thread.yield();

Or call display.sleep() instead. But I do not know when the display is waked up again. (of course you can spawn a thread that calls Display#wake() after the timeout.

Just one question. As far as I remember there was an issue with the completed event and so the improved SWTBrowser injected some JS code into the browser to get notified when the page is really loaded. Is this no longer the case, i.e was is fixed in the Eclipse SWTBrowser ?

Reply:
The issue never appeared during my tests, but I am aware that the possibility exist. For now the SarosApi and the DOM were always available. 
I thought the issue could be related to [BUG 433400](https://bugs.eclipse.org/bugs/show_bug.cgi?id=433400) which was solved by [this commit](https://github.com/eclipse/eclipse.platform.swt/commit/0658d789bfb0cb2ccab2ad6b8af953cadf626c1e).

Reply:
> It seems that this is always called in the SWT EDT. So if you make sure waitForPageLoaded is invoked in the EDT (seems loadUrl is the only method that is not executed on the SWT) you can simply spin wait via.

Thanks for clarifying.